spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchQueryScan.java

```diff
@@ -106,26 +105,42 @@
         }
       }
 
-      if (splitSize != null) {
-        scan = scan.option(TableProperties.SPLIT_SIZE, splitSize.toString());
-      }
-
-      if (splitLookback != null) {
-        scan = scan.option(TableProperties.SPLIT_LOOKBACK, splitLookback.toString());
-      }
-
-      if (splitOpenFileCost != null) {
-        scan = scan.option(TableProperties.SPLIT_OPEN_FILE_COST, splitOpenFileCost.toString());
-      }
+      scan = scan.option(TableProperties.SPLIT_SIZE, String.valueOf(splitSize(scan)));
+      scan = scan.option(TableProperties.SPLIT_LOOKBACK, String.valueOf(splitLookback()));
+      scan = scan.option(TableProperties.SPLIT_OPEN_FILE_COST, String.valueOf(splitOpenFileCost()));
 
       for (Expression filter : filterExpressions()) {
         scan = scan.filter(filter);
       }
 
-      try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
```

Comment:
Quick question, @sunchao. If we exposed a method on `TableScan` like `splitByPartition(boolean)` or any other name and set it from Spark, would we be able to push down this logic into `BaseTableScan`? It seems each query engine would need to implement it if they want to support bucketed joins.

Reply:
Or maybe we even don't need a separate method. We could use whatever we do for other options.

```
if (splitOpenFileCost != null) {
  scan = scan.option(TableProperties.SPLIT_OPEN_FILE_COST, splitOpenFileCost.toString());
}

// use whatever condition we want
if (cond) {
  scan.option("split-by-partition", "true")
}
```

Reply:
Then we could use the same logic in `SparkMergeScan`.

Reply:
Configuring this seems fundamental to the scan, so I would probably make it first class in the scan API. I think you may also want to configure the partition fields or table columns that should be kept separate. That way you'd be able to set `preservePartitionsFor("id")` or something.

Reply:
Thanks. Let me explore how to move this into `BaseTableScan`. One issue I can see is that `SparkMergeScan` currently has its own logic of planning tasks in its `tasks` method, instead of directly calling `TableScan.planTasks`. Therefore we'll need some refactoring in order to avoid duplicating the logic there.